Add JSON format to graph command (#3617)
tuist/tuist: projects/docs/docs/commands/graph.md

The command will output the dependency graph as an image, in the `png` format.
You can also change the format to `dot` (see [DOT](<https://en.wikipedia.org/wiki/DOT_(graph_description_language)>))
to get the raw contents of the graph.
:::caution

> You can also change the format to `json` to export the project structure into a JSON file. The JSON schema is subject

Reviewer: ```suggestion
You can also change the format to `json` to export the graph as a JSON. The JSON schema is subject
```

to change and may be breaking across major versions.

> :::

Reviewer 2: Should be `minor` version? Otherwise I don't see the point of the caution as a major is breaking by definition

Author: Major doesn't necessarily mean breaking though. I'm not sure if we should break the schema in minor versions.

Reviewer: > I'm not sure if we should break the schema in minor versions.

I think we should put out a warning this might break even in minor versions. This enables to iterate fast on this part of the codebase without scaring users by putting out too many major versions.

Author: Ok sounds good!

The command will output the dependency graph as an image, in the `png` format.
You can also change the format to `dot` (see [DOT](<https://en.wikipedia.org/wiki/DOT_(graph_description_language)>))
to get the raw contents of the graph.

> You can also change the format to `json` to export the project structure into a JSON file.

Reviewer 3: Because we are exporting a JSON, I anticipate users building workflows upon that. That means the output becomes a public interface between the users and the CLI. Until the Graph's structure stabilizes, I'd add an [admonition](https://docusaurus.io/docs/markdown-features/admonitions) here making it clear that the format is subject to breaking changes.

Author: Sounds good!

| `kind` | The type of the package. See (Package)[https://docs.tuist.io/manifests/project#package] for all possible values. |
| `path` | Absolute path to the package.                                                                                    |

Sample JSON output:
```
{
    "name": "NAME_OF_MANIFEST",
    "path": "PATH_TO_MANIFEST_DIR",
    "projects": {
        "PATH_TO_PROJECT": {
            "name": "NAME_OF_PROJECT",
            "path": "PATH_TO_PROJECT",
            "targets": [
                {
                    "name": "NAME_OF_TARGET",
                    "product": "TYPE_OF_PRODUCT"
                }
            ],
            "schemes": [
                {
                    "name": "NAME_OF_SCHEME",
                    "testActionTargets": [
                        "TEST_ACTION_TARGET"
                    ]
                }
            ],
            "packages": [
                {
                    "kind": "KIND_OF_PACKAGE_REMOTE_OR_LOCAL",
                    "path": "PATH_OR_URL_OF_PACKAGE"
                }
            ]
        }
    }
}

> ```

Reviewer 2: Are all the paths relative? I would like the generated graph to be the same regardless of the manifest path (hence I would probably remove the top-level path if not needed)

Reviewer: > Are all the paths relative? I would like the generated graph to be the same regardless of the manifest path (hence I would probably remove the top-level path if not needed)

I am not sure if I agree with this. If e.g. all source paths are relative, it will actually make it harder to work with.

Reviewer 2: Why would it make it harder to work it? if they a relative to the path when the command is invoked I don't see a problem

Reviewer: The scenario I am thinking of is going through `sources` to e.g. lint all the files. Also note we are working with strings here and not `RelativePath`, making the conversion a little bit harder.

What would be a good use-case for having relative paths here?

Reviewer 2: Going from relative to absolute should be just a matter of prepending a fixed path, so I don't think it's a big issue.
Any logic you would do on the files, I expect you do it from the project root, so you should be able to use relative paths directly.

I don't have a strong use-case for relative paths, but even being able to compare generated JSONs on different machines could be a nice to have.
I think we should go for relative by default everywhere, unless we have a strong reason to use absolute.

Makes sense? 🚀

Reviewer: I'm fine with that, once we start using this in tasks, we should provide a utility that will convert a string of relative path to a string of absolute path, so it's as easy-as-possible for users to use.

Author: For our use case we actually must have absolute paths. Using relative paths would make our life a lot more painful. Using absolute paths is also consistent with SPM's output. I feel pretty strongly that we should stick with absolute paths here.

Reviewer: Fair! The current format is labeled as unstable, so if we find relative paths would actually be better, I'd revert it. But so far it seems that absolute paths are a better option here.
